File: libnd4j/include/ops/declarable/generic/sparse/dense_to_coo_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// dense_to_coo_bp: backward pass for dense_to_coo.
//
// Forward recap:
//   Input:   [0] dense [rows, cols]  float
//   TArgs:   [0] threshold
//   Outputs: [0] indices [nnz, 2]   INT64
//            [1] values  [nnz]      float
//
// Backward:
//   Inputs:  [0] indices    [nnz, 2]  INT64  (forward output[0])
//            [1] gradValues [nnz]     float  (upstream gradient for values)
//   IArgs:   [0] rows, [1] cols
//   Output:  [0] dDense [rows, cols]  float
//
//   Math: dDense is ZERO-INITIALIZED then for each k in [0, nnz):
//     dDense[ indices[k,0], indices[k,1] ] = gradValues[k]
//   (Scatter with no atomics — dense_to_coo emits unique index pairs.)
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_dense_to_coo_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_coo_bp.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(dense_to_coo_bp, 2, 1, false, 0, 2) {
  auto indices    = INPUT_VARIABLE(0);  // [nnz, 2] INT64
  auto gradValues = INPUT_VARIABLE(1);  // [nnz]    float

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  REQUIRE_TRUE(indices->rankOf() == 2, 0,
               "dense_to_coo_bp: indices must be 2D [nnz, 2], got rank %d",
               indices->rankOf());
  REQUIRE_TRUE(indices->sizeAt(1) == 2, 0,
               "dense_to_coo_bp: indices second dim must be 2, got %lld",
               (long long)indices->sizeAt(1));
  REQUIRE_TRUE(gradValues->rankOf() == 1, 0,
               "dense_to_coo_bp: gradValues must be 1D [nnz], got rank %d",
               gradValues->rankOf());
  REQUIRE_TRUE(indices->sizeAt(0) == gradValues->lengthOf(), 0,
               "dense_to_coo_bp: indices rows (%lld) must match gradValues length (%lld)",
               (long long)indices->sizeAt(0), (long long)gradValues->lengthOf());

  auto dDense = OUTPUT_NULLIFIED(0);  // [rows, cols] float — written by scatter

  sd::ops::helpers::denseToCoo_bp(block.launchContext(),
                                   indices, gradValues, dDense, rows, cols);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(dense_to_coo_bp) {
  // dDense has shape [rows, cols] with the same float dtype as gradValues (input 1).
  auto gradValues = INPUT_VARIABLE(1);
  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(
      gradValues->dataType(), 'c', {rows, cols});

  return SHAPELIST(outShape);
}

DECLARE_TYPES(dense_to_coo_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})    // indices
      ->setAllowedInputTypes(1, {ALL_FLOATS})  // gradValues
      ->setAllowedOutputTypes(0, {ALL_FLOATS}); // dDense
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_dense_to_coo_bp)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/DenseToCooBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Native backward pass for {@link DenseToCoo}: scatters the upstream gradient
 * for the COO values back into a dense gradient matrix.
 *
 * <p>C++ op name: {@code dense_to_coo_bp}
 *
 * <p>Forward recap: {@code dense[rows,cols] → (indices[nnz,2], values[nnz])}.
 *
 * <p><b>Inputs (2 arrays):</b>
 * <ol>
 *   <li>{@code indices}    – 2D [nnz, 2], INT64 — row/col index pairs from the forward pass</li>
 *   <li>{@code gradValues} – 1D [nnz], float — upstream gradient w.r.t. the COO values output</li>
 * </ol>
 *
 * <p><b>Integer arguments (IArgs):</b> {@code rows}, {@code cols}.
 *
 * <p><b>Output (1 array):</b>
 * <ol>
 *   <li>{@code dDense} – 2D [rows, cols], same float dtype as {@code gradValues}</li>
 * </ol>
 *
 * <p>Math: {@code dDense} is zero-initialised, then for each {@code k}:
 * {@code dDense[indices[k,0], indices[k,1]] = gradValues[k]}.
 * No atomics are needed because {@link DenseToCoo} emits unique index pairs.
 *
 * <p>This op is forward-only (no {@code doDiff}); it is the gradient leaf for
 * {@link DenseToCoo#doDiff}.
 */
public class DenseToCooBp extends DynamicCustomOp {

    /** No-arg constructor required for op-registry reflection. */
    public DenseToCooBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param indices    2D [nnz, 2] INT64 — row/col index pairs from the forward pass
     * @param gradValues 1D [nnz] float — upstream gradient w.r.t. the COO values output
     * @param rows       number of rows in the dense matrix
     * @param cols       number of columns in the dense matrix
     */
    public DenseToCooBp(INDArray indices, INDArray gradValues, long rows, long cols) {
        super(new INDArray[]{indices, gradValues}, null);
        addIArgument(rows, cols);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd         the SameDiff graph
     * @param indices    SDVariable [nnz, 2] INT64 — row/col index pairs from the forward pass
     * @param gradValues SDVariable [nnz] float — upstream gradient w.r.t. the COO values output
     * @param rows       number of rows in the dense matrix
     * @param cols       number of columns in the dense matrix
     */
    public DenseToCooBp(SameDiff sd, SDVariable indices, SDVariable gradValues,
                        long rows, long cols) {
        super(sd, new SDVariable[]{indices, gradValues});
        addIArgument(rows, cols);
    }

    @Override
    public String opName() {
        return "dense_to_coo_bp";
    }

    /**
     * Output data type: {@code dDense} has the same float dtype as {@code gradValues}
     * (input index 1).
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Collections.singletonList(dataTypes.get(1));
    }
}
